wallet: introduce fee_rate sat/vB param/option

jonatack opened the pull request:
This PR builds on #11413 and #20220 to address #19543.

- replace overloading the conf_target and estimate_mode params with `fee_rate` in sat/vB in the sendtoaddress, sendmany, send, fundrawtransaction, walletcreatefundedpsbt, and bumpfee RPCs

- allow non-actionable conf_target value of `0` and estimate_mode value of `""` to be passed to use `fee_rate` as a positional argument, in addition to as a named argument

- fix a bug in the experimental send RPC described in https://github.com/bitcoin/bitcoin/pull/20220#discussion_r513789526 where args were not being passed correctly into the options values

- update the feerate error message units for these RPCs from BTC/kB to sat/vB

- update the test coverage, help docs, doxygen docs, and some of the RPC examples

- other changes to address the excellent review feedback

See this wallet meeting log for more context: http://www.erisian.com.au/bitcoin-core-dev/log-2020-11-06.html#l-309

jonatack committed:
wallet: introduce fee_rate (sat/vB) param/option

Create a fee_rate (sat/vB) RPC param and replace overloading the conf_target and
estimate_mode params in the following 6 RPCs with it:

- sendtoaddress
- sendmany
- send
- fundrawtransaction
- walletcreatefundedpsbt
- bumpfee

In RPC bumpfee, the previously existing fee_rate remains but the unit is changed
from BTC/kvB to sat/vB. This is a breaking change, but it should not be an
overly risky one, as the units change by a factor of 1e5 and any fees specified
in BTC/kvB after this commit will either be too low and raise an error or be 1
sat/vB and can be RBFed.

Update the test coverage for each RPC.

Co-authored-by: Murch <murch@murch.one>

jonatack committed:
wallet: remove fee rates from conf_target helps

jonatack committed:
wallet: update sendtoaddress, send RPC examples with fee_rate

jonatack committed:
wallet: fix bug in RPC send options

when empty, options were not being populated by arguments of the same name

found while adding test coverage in 603c0050

jonatack committed:
wallet: remove unneeded WALLET_BTC_KB_TO_SAT_B constant

MarcoFalke commented:
How does this differ/compare to #20250 ?

MarcoFalke commented:
I guess this one is adding a new separate option, #20250 is adding named aliases and keeping the existing options.

jonatack commented:
This PR removes the conf_target/estimate_mode overloading completely and introduces a standard fee rate param.

#20250 keeps the current overloading and makes it more consistent between the six RPCs.

ISTM the question is: do we want the overloading, or a standard feerate param? I'm reasonably sure, from discussing with people, that we don't want the complexity of having both.

jonatack committed:
wallet: update remaining rpcwallet fee rate units to BTC/kvB

jonatack commented:
Changes in latest push:
- In the first commit, "wallet: fix bug in RPC send options", uncommented the previously failing tests and added further tests
- In the second commit, "wallet: remove unneeded WALLET_BTC_KB_TO_SAT_B constant", fixed the patch to ensure all tests continue to pass
- In the remaining commits, made further improvements and fixups seen while re-reviewing per `git diff 078b8fb 2b8d681`

luke-jr commented:
AFAICT, the only "fee_rate" in 0.20 was bumpfee. Specifying BTC/vkB (current value) in place of the new sat/vB would always be a *much* lower feerate. So it'd most likely be too low and error, or worst case lower than you intended and you can just bump it again to fix.

With that in mind, I think the option here should just be renamed to "fee_rate" and break the compatibility for bumpfee.

jonatack commented:
@luke-jr SGTM and that would further simplify the implementation too.

achow101 commented:
Agree with renaming to `fee_rate`. However having two fee rate options for some RPCs is kinda weird. I suppose it's too late to fix that for this release, but it'd be nice to not do that in the future.

Concept ACK.

meshcollider commented:
Concept ACK for `fee_rate`

jonatack committed:
wallet: update fee rate units, use sat/vB for fee_rate error messages

and BTC/kvB for feeRate error messages.

promag reviewed (commented):
Why `fee_rate_sat_vb` and not just `fee_rate`? Also, should not change `verbose` parameter index?

jonatack commented:
> Why `fee_rate_sat_vb` and not just `fee_rate`?

I initially begin with `fee_rate` in #20231. Then, based on https://github.com/bitcoin/bitcoin/pull/20220#discussion_r513789526 I changed to `fee_rate_sat_vb` here.

Today, based on the feedback at last Friday's wallet meeting and here, I will push an update here to use `fee_rate`.

> Also, should not change `verbose` parameter index?

For sendtoaddress and sendmany, I hesitated on this because I thought that users are used to seeing the `verbose` argument placed last, e.g. like gettransaction. If positional `true` or `false` are passed for the `fee_rate`, a type error will be raised and the transaction won't proceed. Is it better to put `fee_rate` after `verbose` instead?

MarcoFalke commented:
Has `verbose` been in a release already?

jonatack committed:
wallet: add fee_rate unit warnings to bumpfee

jonatack commented:
Per d5863c0b3e it looks like `verbose` was added to sendtoaddress and sendmany after the 0.20 release. Not a bugfix, so probably not backported. Checking.

jonatack committed:
wallet: remove redundant bumpfee fee_rate checks

SetFeeEstimateMode() handles these checks now and provides a more actionable
error message.

Xekyo commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4073,26 +4085,31 @@ static RPCHelpMan send()
-                if (!request.params[1].isNull() || !request.params[2].isNull()) {
-                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Use either conf_target and estimate_mode or the options dictionary to control fee rate");
+            UniValue options{request.params[4].isNull() ? UniValue::VOBJ : request.params[4]};
+            if (options.exists("conf_target") || options.exists("estimate_mode") || options.exists("fee_rate")) {
+                if (!request.params[1].isNull() || !request.params[2].isNull() || !request.params[3].isNull()) {
```
Maybe I'm missing something, but I had trouble figuring out from the local code what the parameter set of this particular call is here. If I understand correctly, their types are being defined in 4086-4089. Maybe the parameter names could be added as comments there or here to remind the reader what is being tested to be null here in 4099.

Xekyo replied:
I was trying to figure out what exactly is checked here that prevents e.g. both `conf_target` and `fee_rate` to be set here, if `params[1]…[3]` all need to be null here.

jonatack replied:
Add these comments:
```diff
             RPCTypeCheck(request.params, {
-                UniValueType(), // ARR or OBJ, checked later
-                UniValue::VNUM,
-                UniValue::VSTR,
-                UniValue::VNUM,
-                UniValue::VOBJ,
+                UniValueType(), // outputs (ARR or OBJ, checked later)
+                UniValue::VNUM, // conf_target
+                UniValue::VSTR, // estimate_mode
+                UniValue::VNUM, // fee_rate
+                UniValue::VOBJ, // options
                 }, true
             );
```
If any of these options are passed, then none of the same ones should be passed as arguments. I suppose the logic or error message could be refined further but think I'll punt on that for now. This RPC is still labeled as experimental.

jonatack replied:
> I was trying to figure out what exactly is checked here that prevents e.g. both `conf_target` and `fee_rate` to be set here, if `params[1]…[3]` all need to be null here.

Hm wait, the new error message should be changed. It used to be, "Use either conf_target and estimate_mode or the options dictionary to control fee rate"... will update.

jonatack replied:
Added these comments:
```diff
             RPCTypeCheck(request.params, {
-                UniValueType(), // ARR or OBJ, checked later
-                UniValue::VNUM,
-                UniValue::VSTR,
-                UniValue::VNUM,
-                UniValue::VOBJ,
+                UniValueType(), // outputs (ARR or OBJ, checked later)
+                UniValue::VNUM, // conf_target
+                UniValue::VSTR, // estimate_mode
+                UniValue::VNUM, // fee_rate
+                UniValue::VOBJ, // options
                 }, true
             );
```
> I was trying to figure out what exactly is checked here that prevents e.g. both `conf_target` and `fee_rate` to be set here, if `params[1]…[3]` all need to be null here.

You're right. Updated.
```diff
             UniValue options{request.params[4].isNull() ? UniValue::VOBJ : request.params[4]};
-            if (options.exists("conf_target") || options.exists("estimate_mode") || options.exists("fee_rate")) {
-                if (!request.params[1].isNull() || !request.params[2].isNull() || !request.params[3].isNull()) {
-                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Use either conf_target and estimate_mode for fee estimation, or fee_rate to specify an explicit fee rate.");
+            if (options.exists("conf_target") || options.exists("estimate_mode")) {
+                if (!request.params[1].isNull() || !request.params[2].isNull()) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Pass conf_target and estimate_mode either as arguments or in the options object, but not both");
                 }
             } else {
                 options.pushKV("conf_target", request.params[1]);
                 options.pushKV("estimate_mode", request.params[2]);
+            }
+            if (options.exists("fee_rate")) {
+                if (!request.params[3].isNull()) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Pass fee_rate either as an argument or in the options object, but not both");
+                }
+            } else {
                 options.pushKV("fee_rate", request.params[3]);
             }
```
This hopefully maintains the original goal of the code that either the argument or the option should be passed, but not both. The check if conf_target or estimate_mode are passed at the same time as fee_rate happens in `SetFeeEstimateMode()`.

jonatack replied:
Updated `wallet_send` tests.
```diff
+++ b/test/functional/wallet_send.py
@@ -234,7 +234,7 @@ class WalletSendTest(BitcoinTestFramework):
         for mode in ["unset", "economical", "conservative"]:
             self.test_send(from_wallet=w0, to_wallet=w1, amount=1, arg_conf_target=1, arg_estimate_mode="economical",
                 conf_target=1, estimate_mode=mode, add_to_wallet=False,
-                expect_error=(-8, "Use either conf_target and estimate_mode for fee estimation, or fee_rate to specify an explicit fee rate."))
+                expect_error=(-8, "Pass conf_target and estimate_mode either as arguments or in the options object, but not both"))
 
         self.log.info("Create PSBT from watch-only wallet w3, sign with w2...")
         res = self.test_send(from_wallet=w3, to_wallet=w1, amount=1)
@@ -278,6 +278,10 @@ class WalletSendTest(BitcoinTestFramework):
         fee = self.nodes[1].decodepsbt(res["psbt"])["fee"]
         assert_fee_amount(fee, Decimal(len(res["hex"]) / 2), Decimal("0.00003"))
 
+        # Test that passing fee_rate as both an argument and an option raises.
+        self.test_send(from_wallet=w0, to_wallet=w1, amount=1, arg_fee_rate=1, fee_rate=1, add_to_wallet=False,
+                       expect_error=(-8, "Pass fee_rate either as an argument or in the options object, but not both"))
```

Xekyo replied:
Ooooh, so this was checking that not both the options and arguments were being passed. I guess I hadn't fully grokked that. I guess that would have been fine then, if any of the three was set on one, it would be incorrect for the any of the others to be set in the other parameter set. It actually is a bit more lenient locally then, but since e.g. passing both estimate_mode and fee_rate would throw in other places this is not a problem.

jonatack replied:
Thanks to your feedback, the error messages seem clearer now.

Xekyo commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4073,26 +4085,31 @@ static RPCHelpMan send()
                 throw JSONRPCError(RPC_INVALID_PARAMETER, "Specify estimate_mode");
             }
+            if (options.exists("feeRate")) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Use fee_rate instead of feeRate");
```
Optional: mention that `fee_rate` is specified in `[sat/vB]` other than `feeRate`.

jonatack replied:
Done

jonatack replied:
Done in 136234cb

Xekyo commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -708,74 +707,108 @@ def test_all_watched_funds(self):
+        rawtx = node.createrawtransaction(inputs, outputs)
+
+        result = node.fundrawtransaction(rawtx)  # uses self.min_relay_tx_fee (set by settxfee)
+        result1 = node.fundrawtransaction(rawtx, {"fee_rate": 200000 * self.min_relay_tx_fee})
```
I got hung up on this line because I first thought that `minRelayTxFee` would be 1,000 sats/kvB, but thinking about it, I assume that `self.min_relay_tx_fee` is specified in `BTC/kvB`. I did notice that this does not seem to ever be explicitly stated in this file.

So, the number just sort of comes from nowhere.

jonatack replied:
Updated
```diff
         result = node.fundrawtransaction(rawtx)  # uses self.min_relay_tx_fee (set by settxfee)
-        result1 = node.fundrawtransaction(rawtx, {"fee_rate": 200000 * self.min_relay_tx_fee})
+        btc_kvb_to_sat_vb = 100000  # (1e5)
+        result1 = node.fundrawtransaction(rawtx, {"fee_rate": 2 * btc_kvb_to_sat_vb * self.min_relay_tx_fee})
         result2 = node.fundrawtransaction(rawtx, {"feeRate": 2 * self.min_relay_tx_fee})
-        result3 = node.fundrawtransaction(rawtx, {"fee_rate": 1000000 * self.min_relay_tx_fee})
+        result3 = node.fundrawtransaction(rawtx, {"fee_rate": 10 * btc_kvb_to_sat_vb * self.min_relay_tx_fee})
         result4 = node.fundrawtransaction(rawtx, {"feeRate": 10 * self.min_relay_tx_fee})
```

Xekyo commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -708,74 +707,108 @@ def test_all_watched_funds(self):
-            assert_raises_rpc_error(
-                -8, "Cannot specify both {} and feeRate".format(field),
-                lambda: node.fundrawtransaction(rawtx, {"feeRate": 0.1, field: fee_rate}))
+        # With no arguments passed, expect fee of 141 sats/b.
```
I think this may be a `fee` so the unit should just be `satoshi`, not `sats/B`. Otherwise, I'd not expect the result below to be 10,000×, but rather at a factor of 10,000/141. :thinking:

jonatack replied:
Right! Done.

Xekyo commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -708,74 +707,108 @@ def test_all_watched_funds(self):
-                lambda: node.fundrawtransaction(rawtx, {"feeRate": 0.1, field: fee_rate}))
+        # With no arguments passed, expect fee of 141 sats/b.
+        assert_approx(node.fundrawtransaction(rawtx)["fee"], vexp=0.00000141, vspan=0.00000001)
+        # Expect fee to be 10,000x higher when explicit fee 10,000x greater is specified.
```
This would in turn mean that this sentence should be:

 > Expect fee to be 10,000x higher when explicit **feerate** 10,000x greater is specified.

Also, nit: Wouldn't "10,000× higher" be "10,001 as high"? :nerd_face:

jonatack replied:
Well-spotted :)
```diff
-        # With no arguments passed, expect fee of 141 sats/b.
+        # With no arguments passed, expect fee of 141 satoshis.
         assert_approx(node.fundrawtransaction(rawtx)["fee"], vexp=0.00000141, vspan=0.00000001)
-        # Expect fee to be 10,000x higher when explicit fee 10,000x greater is specified.
+        # Expect fee to be 10,000x higher when explicit fee rate 10,000x greater is specified.
```

Xekyo commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -708,74 +707,108 @@ def test_all_watched_funds(self):
+            node.fundrawtransaction, rawtx, {"fee_rate": 0, "add_inputs": True})
+        assert_raises_rpc_error(-8, "Invalid fee_rate 0.00000000 BTC/kB (must be greater than 0)",
+            node.fundrawtransaction, rawtx, {"feeRate": 0, "add_inputs": True})
+        for param, value in {("fee_rate", 100000), ("feeRate", 1.1)}:
```
I'm not sure if we did that already, but it would be good to assure that the valid value range for `fee_rate` will cause an error for `feeRate` and vice versa. I would expect a lot of people to plug values for either into the other, and I'm not sure off the top of my head that the ranges have no overlap. (They should be shifted by a factor of 1e5 though, right?)

jonatack replied:
Good point. The ranges don't overlap if I'm not confused, except in the minumum fee direction. We already check the min/max for fee_rate (0.1-100,000 sat/vB) and the max for feeRate (1 BTC/kvB) in the functional tests. At the moment, min fee checks with feeRate are disabled due to `coinControl.fOverrideFeeRate = true` in wallet/rpcwallet.cpp `FundTransaction()`.

To test the minimum for feeRate (0.00000999 BTC/kvB), we'd have to remove that `fOverrideFeeRate` setting. I've added this for now:

```diff
test/functional/rpc_fundrawtransaction.py
@@ 772 @@ class RawTransactionsTest(BitcoinTestFramework):
+        # This test will only pass if `coinControl.fOverrideFeeRate = true` in wallet/rpcwallet.cpp::FundTransaction is removed
+        assert_raises_rpc_error(-4, "Fee rate (0.999 sat/vB) is lower than the minimum fee rate setting (1.000 sat/vB)",
+            node.fundrawtransaction, rawtx, {"feeRate": 0.00000999, "add_inputs": True})
```

jonatack replied:
It's not clear to me if `fOverrideFeeRate` should be added to `fee_rate` in fundrawtransaction as well, or just removed. IIUC it currently only serves to not check for a minimum `feeRate` in fundrawtransaction only.

jonatack replied:
It's not clear to me if `fOverrideFeeRate` should be set to true if `fee_rate` is passed in fundrawtransaction or walletcreatefundedpsbt, as it is for `feeRate`, or if the `fOverrideFeeRate` flag could just be removed.

AFAICT `fOverrideFeeRate` set to true currently only serves to bypass minimum feerate checks for `feeRate` in `GetMinimumFeeRate()` for fundrawtransaction and walletcreatefundedpsbt.

jonatack replied:
Found this [comment](https://github.com/bitcoin/bitcoin/pull/10706/#discussion_r126569510) in #10706 that describes the reason in mid-2017 for `fOverrideFeeRate` pertaining to `feeRate` in fundrawtxn (and now wallcreatefundedpsbt).

jonatack replied:
Found this [mini-discussion](https://github.com/bitcoin/bitcoin/pull/10706/#discussion_r126560525) in #10706 in mid-2017 between @ryanofsky and @morcos about the use of `fOverrideFeeRate` for `feeRate` in fundrawtransaction (and now also walletcreatefundedpsbt). Thoughts WRT to using `fOverrideFeeRate` for the new `fee_rate` in these two RPCs -- yay or nay?

jonatack replied:
Decided to maintain the `fOverrideFeeRate` behavior of disabling the min fee checks on `fee_rate` in RPCs fundrawtransaction and walletcreatefundedpsbt only, as has been the case for `feeRate` in these two RPCs since 2017, per the above comments. Done in afe94710ef8. Resolving.

Xekyo commented on test/functional/rpc_psbt.py:
```diff
@@ -187,60 +187,74 @@ def run_test(self):
+        assert_raises_rpc_error(-8, "Invalid fee_rate 0.00000000 BTC/kB (must be greater than 0)",
+            self.nodes[1].walletcreatefundedpsbt, inputs, outputs, 0, {"feeRate": 0, "add_inputs": True})
+        for param, value in {("fee_rate", 100000), ("feeRate", 1.1)}:
+            assert_raises_rpc_error(-4, "Fee exceeds maximum configured by user (e.g. -maxtxfee, maxfeerate)",
```
Probably out of scope, but it would be kinda nice if an error like this told the user what said limit is.

jonatack replied:
Agree--punting on this for the next push but would be good.

Xekyo commented on test/functional/wallet_basic.py:
```diff
@@ -227,65 +230,40 @@ def run_test(self):
-        )
+        self.log.info("Test sendmany with fee_rate param (explicit fee rate in sat/vB)")
+        fee_rate_sat_vb = 2
+        fee_rate_btc_kvb = fee_rate_sat_vb / 100000.0
```
I mean, I see where that 1e-5 is coming from, but if one is not as immersed, it may be a bit more readable, if the two conversions were mentioned more explicitly:

```
fee_rate_btc_kvb = fee_rate_sat_vb * 1e3 / 1e8
```

jonatack replied:
Done
```diff
+++ b/test/functional/wallet_basic.py
@@ -232,7 +232,7 @@ class WalletTest(BitcoinTestFramework):
 
         self.log.info("Test sendmany with fee_rate param (explicit fee rate in sat/vB)")
         fee_rate_sat_vb = 2
-        fee_rate_btc_kvb = fee_rate_sat_vb / 100000.0
+        fee_rate_btc_kvb = fee_rate_sat_vb * 1e3 / 1e8
         explicit_fee_rate_btc_kvb = Decimal(fee_rate_btc_kvb) / 1000
 
         # Passing conf_target 0, estimate_mode "" as placeholder arguments should allow fee_rate to apply.
@@ -405,7 +405,7 @@ class WalletTest(BitcoinTestFramework):
             address = self.nodes[1].getnewaddress()
             amount = 3
             fee_rate_sat_vb = 2
-            fee_rate_btc_kvb = fee_rate_sat_vb / 100000.0
+            fee_rate_btc_kvb = fee_rate_sat_vb * 1e3 / 1e8
```

Xekyo commented on test/functional/wallet_basic.py:
```diff
@@ -227,65 +230,40 @@ def run_test(self):
+        self.log.info("Test sendmany with fee_rate param (explicit fee rate in sat/vB)")
+        fee_rate_sat_vb = 2
+        fee_rate_btc_kvb = fee_rate_sat_vb / 100000.0
+        explicit_fee_rate_vb = Decimal(fee_rate_btc_kvb) / 1000
```
This seems to be the first time we use `BTC/vB`. Given that it's a bit uncommon, that should perhaps be part of the variable name: I suggest to use `explicit_fee_rate_BTC_vb`.

jonatack replied:
Good idea. Renamed to `explicit_fee_rate_btc_kvb` (since the variable names are lowercase).

Xekyo commented on test/functional/wallet_basic.py:
```diff
@@ -420,78 +398,41 @@ def run_test(self):
-
-            self.sync_all(self.nodes[0:3])
+            fee = prebalance - postbalance - Decimal(amount)
+            assert_fee_amount(fee, tx_size, Decimal(fee_rate_sat_vb / 100000.0))
```
Isn't that `Decimal(fee_rate_sat_vb / 100000.0)` just a repetition of `fee_rate_btc_kvb`?

jonatack replied:
Yup, but one tests sendmany and the other tests sendtoaddress. Made this change:
```diff
             fee_rate_sat_vb = 2
+            fee_rate_btc_kvb = fee_rate_sat_vb / 100000.0
.../...
             fee = prebalance - postbalance - Decimal(amount)
-            assert_fee_amount(fee, tx_size, Decimal(fee_rate_sat_vb / 100000.0))
+            assert_fee_amount(fee, tx_size, Decimal(fee_rate_btc_kvb))
```

Xekyo commented on test/functional/wallet_basic.py:
```diff
@@ -227,65 +230,40 @@ def run_test(self):
+        fee_rate_sat_vb = 2
+        fee_rate_btc_kvb = fee_rate_sat_vb / 100000.0
+        explicit_fee_rate_vb = Decimal(fee_rate_btc_kvb) / 1000
+        # Passing conf_target 0, estimate_mode "" as placeholder arguments should allow fee_rate to apply.
```
Okay, this is as far as I got in 53b35be8f87fd894614a5e3609b1d8895a96c93c. I require nourishment. As discussed will continue after lunch.

jonatack replied:
Thanks for the good ideas and feedback, @Xekyo. I'll try to implement them (and any other feedback you may have while continuing) tomorrow in the morning to move this forward. Heading to bed :)

murchandamus reviewed (commented):
Partial review up to ~3/4 of 53b35be8f87fd894614a5e3609b1d8895a96c93c. Will continue in approx 60 minutes.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3120,15 +3122,21 @@ void FundTransaction(CWallet* const pwallet, CMutableTransaction& tx, CAmount& f
-        {
+        if (options.exists("feeRate")) {
+            if (options.exists("fee_rate")) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Cannot specify both fee_rate (" + CURRENCY_ATOM + "/vB) and feeRate (" + CURRENCY_UNIT + "/kvB)");
```
Yes, reckon we should deprecate `feeRate` as soon as feasible after this.

jonatack replied:
> Yes, reckon we should deprecate `feeRate` as soon as feasible after this.

Done in #20483

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -217,12 +215,8 @@ static void SetFeeEstimateMode(const CWallet* pwallet, CCoinControl& cc, const U
-
-        cc.m_feerate = CFeeRate(fee_rate);
+        CAmount feerate{AmountFromValue(estimate_param)};
+        cc.m_feerate = cc.m_fee_mode == FeeEstimateMode::SAT_B ? CFeeRate(feerate, COIN) : CFeeRate(feerate);
```
Yes :+1: and I was more confident about making this change after all the test coverage we added in #20220.

achow101 replied:
It would be nice to have a comment explaining this because it is not immediately clear why this works and future work on this code may not see this explanation.

jonatack replied:
Done, added doxygen documentation to the CFeeRate ctor.

jonatack replied:
Done, added doxygen documentation.

```diff
-    /** Constructor for a fee rate in satoshis per kB. The size in bytes must not exceed (2^63 - 1)*/
+
+    /** Constructor for a fee rate in satoshis per kvB (sat/kvB). The size in bytes must not exceed (2^63 - 1).
+     *
+     *  Passing an nBytes value of COIN (1e8) returns a fee rate in satoshis per vB (sat/vB),
+     *  e.g. (nFeePaid * 1e8 / 1e3) == (nFeePaid / 1e5),
+     *  where 1e5 is the ratio to convert from BTC/kvB to sat/vB.
+     *
+     *  @param[in] nFeePaid  CAmount fee rate to construct with
+     *  @param[in] nBytes    size_t units to construct with, if > 0
+     *  @returns   fee rate
+     */
     CFeeRate(const CAmount& nFeePaid, size_t nBytes);
```

jonatack replied:
Done, added Doxygen documentation in a3eac6e603fddeda1dae20fba840d42f02710531

jonatack replied:
I asked myself why else the two-argument ctor might have been written and didn't come up with an answer...I'm not against reviewing a refactoring of it.

jonatack replied:
Looking at settxfee in `wallet/rpcwallet.cpp`, this method is used in a similar manner.
```
    CAmount nAmount = AmountFromValue(request.params[0]);
    CFeeRate tx_fee_rate(nAmount, 1000);
```

jonatack replied:
> This feels like a hack, unless I misunderstand the original purpose of this constructor. Maybe it's better to add a separate constructor with a boolean to distinguish `SAT_VB` from `BTC/kvB`.

@Sjors done, proposed a separate constructor in https://github.com/bitcoin/bitcoin/commit/9c479bfc293fec9063acb9667ddb3cb60b46469d

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4087,7 +4087,7 @@ static RPCHelpMan send()
             CWallet* const pwallet = wallet.get();
 
-            UniValue options = request.params[3];
+            UniValue options{request.params[3].isNull() ? UniValue::VOBJ : request.params[3]};
```
Yes, that was my read on this too. And if `options` is empty, it is populated from the arg params a few lines after:
```cpp
            } else {
                options.pushKV("conf_target", request.params[1]);
                options.pushKV("estimate_mode", request.params[2]);
                options.pushKV("fee_rate", request.params[3]);
            }
```

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -198,30 +198,29 @@ static std::string LabelFromValue(const UniValue& value)
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid estimate_mode parameter");
+    if (!fee_rate.isNull()) {
+        if (!conf_target.isNull() && conf_target.get_int() > 0) { // conf_target value of 0 allowed as positional placeholder
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Cannot specify both conf_target and fee_rate. Please provide either a confirmation target in blocks for automatic fee estimation, or an explicit fee rate.");
```
This error message will only be raised/returned if a valid conf_target value is passed, not a placeholder value of `0`.

The positional order depends on the RPC, so the user would naturally need to refer to the specific RPC help. I tried to generally keep to the following positional order, where possible: conf_target, estimate_mode, fee_rate.

Xekyo commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4058,9 +4063,15 @@ static RPCHelpMan send()
+        + HelpExampleCli("send", "'{\"" + EXAMPLE_ADDRESS[0] + "\": 0.2}' '{\"fee_rate\": 1}'\n") +
+        "Send 0.3 BTC with a fee rate of 25 " + CURRENCY_ATOM + "/vB using named arguments\n"
+        + HelpExampleCli("-named send", "outputs='{\"" + EXAMPLE_ADDRESS[0] + "\": 0.3}' fee_rate=25\n") +
+        "Create a transaction that should confirm the next block, with a specific input, and return result without adding to wallet or broadcasting to the network\n"
```
Thanks for updating all these! I think they hadn't been filled-in since 2013ish given the above mention of Sean's Outpost :)

murchandamus reviewed (commented):
Looks great to me! Gonna circle back tomorrow after I get up. :)

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -454,12 +454,17 @@ static RPCHelpMan sendtoaddress()
+                    "\nSend 0.1 BTC\n"
+                    + HelpExampleCli("sendtoaddress", "\"" + EXAMPLE_ADDRESS[0] + "\" 0.1") +
+                    "\nSend 0.1 BTC with a confirmation target of 6 blocks in economical fee estimate mode using positional arguments\n"
+                    + HelpExampleCli("sendtoaddress", "\"" + EXAMPLE_ADDRESS[0] + "\" 0.1 \"lunch\" \"seans outpost\" false true 6 economical") +
```
TIL, thanks! Fixing

jonatack commented on test/functional/wallet_send.py:
```diff
@@ -282,17 +285,15 @@ def run_test(self):
         for mode in ["foo", Decimal("3.141592")]:
```
Good idea, added.
```diff
-        for mode in ["foo", Decimal("3.141592")]:
+        for mode in ["", "foo", Decimal("3.141592")]:
```

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -198,30 +198,29 @@ static std::string LabelFromValue(const UniValue& value)
- * @param[in]     estimate_mode  String value (e.g. "ECONOMICAL")
- * @param[in]     estimate_param Parameter (blocks to confirm, explicit fee rate, etc)
- * @throws a JSONRPCError if estimate_mode is unknown, or if estimate_param is missing when required
+ * @param[in]     conf_target    UniValue integer, confirmation target in blocks
```
Good idea, addressed this and https://github.com/bitcoin/bitcoin/pull/20305#discussion_r519998042 with
```diff
@@ -198,17 +198,20 @@ static std::string LabelFromValue(const UniValue& value)
  *
  * @param[in]     pwallet        Wallet pointer
  * @param[in,out] cc             Coin control which is to be updated
- * @param[in]     conf_target    UniValue integer, confirmation target in blocks
- * @param[in]     estimate_mode  UniValue string, fee estimation mode ("unset", "economical" or "conservative")
- * @param[in]     fee_rate       UniValue real, fee rate in sat/vB
+ * @param[in]     conf_target    UniValue integer, confirmation target in blocks, values between 1 and 1008 are valid per policy/fees.h;
+ *                                   0 is allowed as a no-op positional placeholder if a fee_rate is passed
+ * @param[in]     estimate_mode  UniValue string, fee estimation mode, valid values are "unset", "economical" or "conservative";
+ *                                   "" is allowed as a no-op positional placeholder if a fee_rate is passed
+ * @param[in]     fee_rate       UniValue real, fee rate in sat/vB; if a fee_rate is present then both
+ *                                   conf_target and estimate_mode must either be null or no-op values
  */
 static void SetFeeEstimateMode(const CWallet* pwallet, CCoinControl& cc, const UniValue& conf_target, const UniValue& estimate_mode, const UniValue& fee_rate)
 {
     if (!fee_rate.isNull()) {
-        if (!conf_target.isNull() && conf_target.get_int() > 0) { // conf_target value of 0 allowed as positional placeholder
+        if (!conf_target.isNull() && conf_target.get_int() > 0) {
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Cannot specify both conf_target and fee_rate. Please provide either a confirmation target in blocks for automatic fee estimation, or an explicit fee rate.");
         }
-        if (!estimate_mode.isNull() && !estimate_mode.get_str().empty()) { // estimate_mode value of "" allowed as positional placeholder
+        if (!estimate_mode.isNull() && !estimate_mode.get_str().empty()) {
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Cannot specify both estimate_mode and fee_rate");
         }
         cc.m_feerate = CFeeRate(AmountFromValue(fee_rate), COIN);
```

Xekyo replied:
Excellent, I like this very thorough comment.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -198,30 +198,29 @@ static std::string LabelFromValue(const UniValue& value)
- * @throws a JSONRPCError if estimate_mode is unknown, or if estimate_param is missing when required
+ * @param[in]     conf_target    UniValue integer, confirmation target in blocks
+ * @param[in]     estimate_mode  UniValue string, fee estimation mode ("unset", "economical" or "conservative")
+ * @param[in]     fee_rate       UniValue real, fee rate in sat/vB
```
:+1: done as described in https://github.com/bitcoin/bitcoin/pull/20305#discussion_r520414437

jonatack committed:
wallet: provide valid values if invalid estimate mode passed

Co-authored-by: Murch <murch@murch.one>

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -198,30 +198,29 @@ static std::string LabelFromValue(const UniValue& value)
+        return;
+    }
+    if (!estimate_mode.isNull() && !FeeModeFromString(estimate_mode.get_str(), cc.m_fee_mode)) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid estimate_mode parameter");
```
Good idea, doing.

jonatack replied:
> If I were a user, I would like to get instructions about what the valid parameters are, if that's not too much work.

Done in b827d2dbc229a

jonatack committed:
wallet: use MIN_RELAY_TX_FEE in bumpfee help

Co-authored-by: Murch <murch@murch.one>

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3386,16 +3396,16 @@ static RPCHelpMan bumpfee_helper(std::string method_name)
-                             "Must be at least 0.0001 " + CURRENCY_UNIT + "/kB higher than the current transaction fee rate.\n"},
+                    {"fee_rate", RPCArg::Type::AMOUNT, /* default */ "not set, fall back to wallet fee estimation",
+                             "\nSpecify a fee rate in " + CURRENCY_ATOM + "/vB instead of relying on the built-in fee estimator.\n"
+                             "Must be at least 1 " + CURRENCY_ATOM + "/vB higher than the current transaction fee rate.\n"},
```
Done, will add the commit hash after pushing.

jonatack replied:
Done in 802193e73e22580ec77

jonatack commented on test/functional/wallet_bumpfee.py:
```diff
@@ -105,50 +103,43 @@ def test_invalid_parameters(self, rbf_node, peer_node, dest_address):
-        for options in [{"fee_rate": INSUFFICIENT}, {"conf_target": INSUFFICIENT, "estimate_mode": BTC_MODE}, {"conf_target": 1, "estimate_mode": SAT_MODE}]:
-            assert_raises_rpc_error(-8, "Insufficient total fee", rbf_node.bumpfee, rbfid, options)
+        # Bumping to just above minrelay should fail to increase the total fee enough.
+        assert_raises_rpc_error(-8, "Insufficient total fee 0.00000141, must be at least 0.00001704 (oldFee 0.00000999 + incrementalFee 0.00000705)",
```
I think that's from `wallet/wallet.h`, though the different fee constants and config options are a bit confusing to me.
```
//! minimum recommended increment for BIP 125 replacement txs
static const CAmount WALLET_INCREMENTAL_RELAY_FEE = 5000;
```

Xekyo replied:
Thanks for tracking that down!

jonatack commented on test/functional/wallet_bumpfee.py:
```diff
@@ -105,50 +103,43 @@ def test_invalid_parameters(self, rbf_node, peer_node, dest_address):
+        self.log.info("Test invalid fee rate settings")
+        assert_raises_rpc_error(-8, "Invalid fee_rate 0.00000000 BTC/kB (must be greater than 0)",
+            rbf_node.bumpfee, rbfid, {"fee_rate": 0})
+        assert_raises_rpc_error(-4, "Specified or calculated fee 0.141 is too high (cannot be higher than -maxtxfee 0.10",
```
Could be a good follow-up.

Sjors replied:
I proposed that once: #16539

jonatack replied:
> woah, damn. I think we may want to have a look at the `maxtxfee` that's a lot of money.

@Xekyo TIL, as @Sjors mentions below, this was proposed in https://github.com/bitcoin/bitcoin/pull/16539

jonatack replied:
> I proposed that once: #16539

TIL, thanks. Interesting discussion in that PR.

jonatack commented on src/policy/feerate.h:
```diff
@@ -19,8 +19,8 @@ enum class FeeEstimateMode {
     CONSERVATIVE, //!< Force estimateSmartFee to use conservative estimates
-    BTC_KB,       //!< Use explicit BTC/kB fee given in coin control
-    SAT_B,        //!< Use explicit sat/B fee given in coin control
+    BTC_KVB,      //!< Use BTC/kvB fee rate unit
+    SAT_VB,       //!< Use sat/vB fee rate unit
```
It's still the default unit elsewhere (until the remaining RPCs move to sat/vB). For now, this PR remains in the scope of the 6 RPCs described in the PR description and the "wallet: introduce fee_rate (sat/vB) param/option" commit message. I was indeed tempted to scope-creep it to include some of the other RPCs :)

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2303,7 +2303,7 @@ static RPCHelpMan settxfee()
                 "Can be deactivated by passing 0 as the fee. In that case automatic fee selection will be used by default.\n",
                 {
-                    {"amount", RPCArg::Type::AMOUNT, RPCArg::Optional::NO, "The transaction fee in " + CURRENCY_UNIT + "/kB"},
+                    {"amount", RPCArg::Type::AMOUNT, RPCArg::Optional::NO, "The transaction fee in " + CURRENCY_UNIT + "/kvB"},
```
Yep, scope creep on an already-wide PR. Could be a next step if you want to tackle it.

jonatack replied:
Done in #20391.

achow101 commented:
ACK 5660dd301152c0c7a70934cc7650ad162a44c9b4

jonatack committed:
wallet: add CFeeRate ctor doxygen documentation

as requested by reviewers

jonatack commented:
Thanks for all the feedback. Re-checking all the fiddly things, will push the update in a few minutes.

jonatack commented:
Addressed @Xekyo and @achow101 feedback (thanks!), fixed a few things seen while self-reviewing, and removed a few duplicate tests/added some missing tests in `rpc_fundrawtransaction.py` per `git diff 5660dd3 bcb8e0d`

jonatack commented:
Closed and reopened to see if the valgrind ci run will stop timing out.

jonatack commented:
Rebased on master after merge of #20368 to hopefully avoid the valgrind fuzz CI job time-outs.

No other change other than renaming and moving the CFeeRate doxygen commit to just before "wallet: remove unneeded WALLET_BTC_KB_TO_SAT_B constant" for clarity, diff per `git range-diff 155bf91 bcb8e0d 94479c8`

jonatack committed:
wallet: override minfee checks (fOverrideFeeRate) for fee_rate

in RPCs fundrawtransaction and walletcreatefundedpsbt only.

This provides the existing feeRate (BTC/kvB) behavior in these two RPCs to the
new fee_rate (sat/vB) param also.

See these two GitHub review discussions for more info:
https://github.com/bitcoin/bitcoin/pull/10706/#discussion_r126560525
https://github.com/bitcoin/bitcoin/pull/20305#discussion_r520032533

achow101 commented:
re-ACK 05e82d8

Sjors reviewed (approved):
Concept ACK

utACK 05e82d86b09d914ebce05dbc92a7299cb026847b

Concept meh on adding the `v` (`BTC/kvB` instead of `BTC/kB`, `sat/vB` instead of `sat/B`). The distinction between bytes and "virtual" bytes has no practical use. It's only interesting for people who want to understand the internals of Bitcoin or why spending from a SegWit is address is cheaper.

I found the overloaded `estimate_mode` param more intuitive as a `bitcoin-cli` user, i.e. `send ... 1 sat/b` before vs. `send ... 0 "" 1` now. But the example in the help explains how to use it. Since the `send` RPC is experimental, we can overhaul it in a future version anyway.

I agree that the bumpfee unit change seems safe, even when users downgrade their node, as explained in the commit message of a0d495747320c79b27a83c216dcc526ac8df8f24, except if you downgrade to a version before #16257 (i.e. before v0.19, and the fix has also been backported).

Do we also want to do something about the `unset` estimate mode? Not sure what it does. It seems safest to make `economical` the default estimate mode. But this PR already has a very large scope.

jonatack commented:
> Concept meh on adding the `v` (`BTC/kvB` instead of `BTC/kB`, `sat/vB` instead of `sat/B`). The distinction between bytes and "virtual" bytes has no practical use. It's only interesting for people who want to understand the internals of Bitcoin or why spending from a SegWit is address is cheaper.

I understand and don't have a strong opinion on sat/B vs sat/vB, happy to follow review consensus.

MarcoFalke commented:
Concept ACK 05e82d86b0

I conceptually agree with the changes, but I haven't reviewed them

Xekyo commented:
> Concept meh on adding the `v` (`BTC/kvB` instead of `BTC/kB`, `sat/vB` instead of `sat/B`). The distinction between bytes and "virtual" bytes has no practical use. It's only interesting for people who want to understand the internals of Bitcoin or why spending from a SegWit is address is cheaper.

I dissent: a bunch of blockexplorers still show "feerate" in `sat/B`, but it often refers to "fee per byte length" of the transaction (ignoring weight), some really outdated blockexplorers have until recently shown fee per stripped size under that label. In my experience, some users were confused by that, and tried to e.g. CPFP a transaction to an amount slightly higher than the "fee per byte length" which happened to still be below the actual feerate. Explicitly stating the feerate with the actually relevant measure potentially resolves that sort of confusion. Maybe that would even be a good argument to fully switch to `satoshi/weightunit`, but here we are.

Xekyo commented:
tACK 05e82d86b09d914ebce05dbc92a7299cb026847b

Sjors commented:
If there's still block explorers (that people use) using real bytes instead of virtual bytes in their fee calculation, then indeed adding the `v` might be necessary. Either way, don't let it hold back this PR.

Xekyo commented:
Just to provide two examples: 

This first is from Blockchair providing the fee per raw size as [sat/B] next to the actual feerate: ![image](https://user-images.githubusercontent.com/4060799/99264847-fcafa000-27ee-11eb-86d1-2f3cfb7b66b6.png)

This second is Blockchain.com, providing fee per raw size as [sat/B] next to actual [sat/WU]:
![image](https://user-images.githubusercontent.com/4060799/99265376-a131e200-27ef-11eb-890f-d8d693d48f61.png)

luke-jr commented:
Block explorer websites have been confusing users with bogus "information" forever. Trying to match them doesn't make sense.

MarcoFalke commented:
review ACK 05e82d86b0 did not test and found a few style nits, which can be fixed later 🍯

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 05e82d86b0 did not test and found a few style nits, which can be fixed later 🍯
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUi04QwAtjdQi2xSn12WwCON4MXuHDxB7owu5U1QxHMD5UpoaASOromS45wnAaaS
X2VhwFoYWJWkPAeCycZ0yr+tyhjr91Z8JqcPpPheryzIKyVys1QMSVVU4Nc4m/2D
PrQOhzn5BX74CnYU+lNoqRECzKslJGXypjWnJz8zmyGU41IDbl7LxQf5sOwYw0kD
k0LDLAcQyg1jLhAwlIvhFMccvzUGz49eXxRfpSO/8qghRm4thKyLvQ7l3nN+zGM7
lQuYdo3MU+OOi82fHJAnsLVwziDVxe0hVr1/r4OHaaeRMJsFSzPVRVox/JV7Zq1E
B8OG5GtifbGR1TqTYhUFZaNCVLKiFCmiKn2SY1VR6UoGonHGgtz/JnQzmpSJbqAd
5i4HtojlkPgtMlLuEadZWmwVNNcpJzgpWjubbc0489/5rrhBg9E+tSSQ0KvclzTg
ZBAEKdSJWy/jGvyxZri7Rzp+Z8c13cLo/oWKw98Y6PV12+uNNv2M+DV/jdQkEasW
GaGTrQnL
=h9+z
-----END PGP SIGNATURE-----
```

python: error while loading shared libraries: libpython3.8.so.1.0: cannot open shared object file: No such file or directory

</details>

Sjors commented:
I contacted one such explorer and nagged them to switch from bytes to vbytes :-)

MarcoFalke commented:
This probably needs release notes?

MarcoFalke commented:
Btw, the first commit adds tests that pass without the code changes. I haven't tried the other added tests, but as a general rule, I'd prefer if new tests were added in a separate commit and not in the same commit that also changes behaviour.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3373,35 +3392,38 @@ RPCHelpMan signrawtransactionwithwallet()
 static RPCHelpMan bumpfee_helper(std::string method_name)
 {
     bool want_psbt = method_name == "psbtbumpfee";
+    const std::string incremental_fee{CFeeRate(DEFAULT_MIN_RELAY_TX_FEE).ToString(FeeEstimateMode::SAT_VB)};
```
DEFAULT_INCREMENTAL_RELAY_FEE, no?

jonatack replied:
Agreed, this looks like an oversight.

jonatack replied:
`#include <validation.h>` -> `#include <policy.h>` in rpcwallet.cpp with the change

jonatack replied:
Done in #20426, thanks for the catch @MarcoFalke

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3120,15 +3122,21 @@ void FundTransaction(CWallet* const pwallet, CMutableTransaction& tx, CAmount& f
             }
-            coinControl.m_feerate = CFeeRate(AmountFromValue(options["feeRate"]));
+            CFeeRate fee_rate(AmountFromValue(options["feeRate"]));
+            if (fee_rate <= CFeeRate(0)) {
```
This is a regression. In 0.20 it is possible to fund a raw tx without fee

jonatack replied:
Should still be possible, will look at it today

jonatack replied:
It is possible to call fundrawtransaction (or walletcreatefundedpsbt) without passing a feeRate or a fee_rate, as they are optional args.

Behavior before a0d495747320c79b27a83c216dcc526ac8df8f24: a negative feeRate raised "Amount out of range", and passing a feeRate of zero was possible in these two calls, but not in bumpfee or send{toaddress, many}.

Current behavior is the same except that the > 0 feeRate check was extended to fundraw and WCFP.

I will remove the zero check on feeRate/fee_rate for those two calls.

jonatack replied:
Discussing offline with @harding and @Xekyo, zero-fee valid txns used to be a thing and might make sense in some future situations (package relay, LN anchor commitments) even if they are non-standard.

MarcoFalke replied:
"zero-fee" txs could also pay fees externally (out-of-band), not even in the same package

jonatack replied:
Done in #20426

jonatack commented:
> This probably needs release notes?

Yes, will update the wiki.

MarcoFalke commented:
Turns out the style nits are regressions. See https://github.com/bitcoin/bitcoin/pull/20305/files#r525406176 and #20410

MarcoFalke commented on test/functional/wallet_basic.py:
```diff
@@ -420,78 +399,43 @@ def run_test(self):
+            # Passing conf_target 0, estimate_mode "" as placeholder arguments should allow fee_rate to apply.
+            txid = self.nodes[2].sendtoaddress(address=address, amount=amount, conf_target=0, estimate_mode="", fee_rate=fee_rate_sat_vb)
             tx_size = self.get_vsize(self.nodes[2].gettransaction(txid)['hex'])
-            self.sync_all(self.nodes[0:3])
```
Why is this safe to remove?

MONIMAKER365 reviewed (commented):
#18261

jonatack commented:
Release notes added to the wiki at https://github.com/bitcoin-core/bitcoin-devwiki/wiki/0.21.0-Release-Notes-Draft.

The `Needs release note` tag can be removed.
